Sources/NIOSSH/SshAgent/Protocol.swift

```
@@ -0,0 +1,168 @@
+
+        switch self {
+        case .signRequest(keyBlob: let blob, let data, let flags):
+            buf.writeInteger(UInt32(blob.count))
+            buf.writeBytes(blob)
+            buf.writeInteger(UInt32(data.count))
+            buf.writeBytes(data)
+            buf.writeInteger(flags)
+        case .addIdentity(let id):
+            for var bb in id.identity {
+                buf.writeSSHString(&bb)
+            }
+        default:
+            break
+        }
+    }
+}
+
+/// An identifier representing an Agent stored identity
+///
+/// The SSH Agent doesn't allow direct access to private keys
+/// so this type represents the public identity and is used
+/// as a key to the SSH Agent when performing operations
+public struct SshIdentity: Hashable, Sendable {
+    let key: ByteBuffer
+    let comment: String
+
+    public var keyBlob: [UInt8] {
+        key.getBytes(at: 0, length: key.readableBytes) ?? []
+    }
+}
+
+extension SshIdentity: CustomStringConvertible {
+    public var description: String {
+        var bb = ByteBuffer(buffer: key)
+        guard
+            let id = bb.readSSHStringAsString(),
+            let bytes = bb.readSSHString()
+        else {
+            return "unknown \(key) \(comment)"
+        }
+
+        return "\(id) \(Data(bytes.readableBytesView).base64EncodedString()) \(comment)"
+    }
+}
+
+private func readIdentityList(_ buf: inout ByteBuffer) -> [SshIdentity] {
+    guard let nKeys: UInt32 = buf.readInteger() else {
+        return []
+    }
+    var result: [SshIdentity] = []
+    result.reserveCapacity(Int(nKeys))
+
+    for _ in 0..<nKeys {
+        guard
+            let key = buf.readSSHString(),
+            let comment = buf.readSSHStringAsString()
+        else {
+            return result
+        }
+        result.append(SshIdentity(key: key, comment: comment))
+    }
+
+    if buf.readableBytes > 0 {
+        print("Bytes left after reading list \(buf)")
```

This should probably throw an error, not print.